velox/common/base/PeriodicStatsReporter.cpp

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/common/base/PeriodicStatsReporter.h"
+#include "velox/common/base/Counters.h"
+#include "velox/common/base/StatsReporter.h"
+#include "velox/common/caching/AsyncDataCache.h"
+#include "velox/common/caching/CacheTTLController.h"
+#include "velox/common/memory/Memory.h"
+#include "velox/common/memory/MmapAllocator.h"
+
+namespace facebook::velox {
+
+namespace {
+#define REPORT_IF_NOT_ZERO(name, counter)   \
+  if ((counter) != 0) {                     \
+    RECORD_METRIC_VALUE((name), (counter)); \
+  }
+} // namespace
+
+PeriodicStatsReporter::PeriodicStatsReporter(
+    const velox::memory::MemoryAllocator* allocator,
+    const velox::cache::AsyncDataCache* cache,
+    const velox::memory::MemoryArbitrator* arbitrator,
+    Options options)
+    : allocator_(allocator),
+      cache_(cache),
+      arbitrator_(arbitrator),
+      options_(options) {
+  lastCacheStats_.ssdStats = std::make_shared<cache::SsdCacheStats>();
```

Comment:
Why not explicit check if ssd is null in stats update?

Reply:
we can do that, too